src/Middleware/RequestLimiter/src/RequestLimiterMiddleware.cs

```diff
@@ -0,0 +1,118 @@
+            var attributes = endpoint?.Metadata.GetOrderedMetadata<RequestLimitAttribute>();
+
+            if (attributes == null)
+            {
+                await _next.Invoke(context);
+                return;
+            }
+
+            var resourceLeases = new Stack<ResourceLease>();
+            try
+            {
+                foreach (var attribute in attributes)
+                {
+                    if (!string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter != null)
+                    {
+                        throw new InvalidOperationException("Cannot specify both policy and limiter");
+                    }
+
+                    if (string.IsNullOrEmpty(attribute.Policy) && attribute.Limiter == null)
+                    {
+                        if (_options.ResolveDefaultRequestLimit != null)
+                        {
+                            if (!await ApplyLimitAsync(_options.ResolveDefaultRequestLimit(context.RequestServices), context, resourceLeases))
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    // Policy based limiters
+                    if (!string.IsNullOrEmpty(attribute.Policy))
+                    {
+                        if (!_options.PolicyMap.TryGetValue(attribute.Policy, out var policy))
+                        {
+                            throw new InvalidOperationException("Policy not found");
+                        }
+
+                        foreach (var limitResolver in policy.LimiterResolvers)
+                        {
+                            if (!await ApplyLimitAsync(limitResolver(context.RequestServices), context, resourceLeases))
+                            {
+                                return;
+                            }
+                        }
+                    }
+
+                    if (attribute.Limiter != null)
+                    {
+                        // Registrations based limiters
+                        if (!await ApplyLimitAsync(attribute.Limiter, context, resourceLeases))
+                        {
+                            return;
+                        }
+                    }
+
+                }
+
+                await _next.Invoke(context);
+            }
+            finally
+            {
+                while (resourceLeases.TryPop(out var resource))
+                {
+                    _logger.LogInformation("Releasing resource");
+                    resource.Dispose();
+                }
+            };
+        }
+
+        private async Task<bool> ApplyLimitAsync(AggregatedResourceLimiter<HttpContext> limiter, HttpContext context, Stack<ResourceLease> obtainedResources)
+        {
+            _logger.LogInformation("Resource count: " + limiter.EstimatedCount(context));
+            var resourceLease = await limiter.WaitAsync(context);
+            if (!resourceLease.IsAcquired)
```

Comment:
Add a fast path to avoid the async/await for the sync scenario. Most limiters will complete immediately until there's pressure, right?

Reply:
@Tratcher got the perf bug!

Reply:
👍